File: T-Bank/Algorithms-and-Data-Structures-2025/1sprint-IntroductionToSorting-BinarySearch/contest/D-Квадратный_корень_и_квадратный_квадрат/solution.py
import sys

PRECISE: float = 0.000001

def solution() -> None:
    """
    Time Complexity O(logN)
    Memory Complexity O(1)
    """
    c: float = float(sys.stdin.readline().rstrip())
    left: float = 0
    right: float = c
    while abs(left - right) != 0:
        middle: float = left + (right - left) / 2
        c_prime: float = middle**2 + (middle + 1)**.5
        if abs(c - c_prime) <= PRECISE:
            sys.stdout.write(f'{middle}\n')
            return
        elif c_prime > c: right = middle
        else: left = middle


if __name__ == '__main__':
    solution()

File: T-Bank/Algorithms-and-Data-Structures-2025/1sprint-IntroductionToSorting-BinarySearch/contest/D-Квадратный_корень_и_квадратный_квадрат/solution.cpp
#include <iostream>
#include <cmath>
#include <iomanip>


void solution() {
    /*
    Time Complexity O(logN)
    Memory Complexity O(1)
    */
    double PRECISE = 0.000001;
    double c;
    std::cin >> c;
    double left = 0;
    double right = c;
    while (std::abs(left - right) > 0.0) {
        double middle = left + (right - left) / 2.0;
        double cPrime = std::pow(middle, 2) + std::sqrt(middle + 1.0);
        if (std::abs(cPrime - c) <= PRECISE) {
            std::cout << std::fixed << std::setprecision(6) << middle << "\n";
            return;
        } else if (cPrime > c) right = middle;
        else left = middle;
    }
}


int main() {
    solution();
}